components/builder-depot/src/server.rs

```diff
@@ -2307,6 +2264,393 @@ fn target_from_headers(user_agent_header: &UserAgent) -> result::Result<PackageT
+        // Don't schedule builds if the upload is being done by the builder
+        if config.builds_enabled && (ident.get_origin() == "core" || config.non_core_builds_enabled)
+            && builder_flag.is_none()
+        {
+            if config.jobsrv_enabled {
+                let mut request = JobGroupSpec::new();
+                request.set_origin(ident.get_origin().to_string());
+                request.set_package(ident.get_name().to_string());
+                request.set_target(target.to_string());
+                request.set_deps_only(true);
+                request.set_origin_only(!config.non_core_builds_enabled);
+                request.set_package_only(false);
+                request.set_trigger(JobGroupTrigger::Upload);
+                request.set_requester_id(owner_id);
+                request.set_requester_name(owner_name);
+
+                match conn.route::<JobGroupSpec, JobGroup>(&request) {
+                    Ok(group) => debug!(
+                        "Scheduled reverse dependecy build for {}, group id: {}, origin_only: {}",
+                        ident,
+                        group.get_id(),
+                        !config.non_core_builds_enabled
+                    ),
+                    Err(err) => warn!("Unable to schedule build, err: {:?}", err),
+                }
+            }
+        }
+    }
+
+    Ok(())
+}
+
+// This function is called from a background thread, so we can't pass the Request object into it.
+fn download_package_from_upstream_depot(
+    depot: DepotUtil,
+    depot_cli: DepotClient,
+    oauth_token: Option<String>,
+    upstream_depot: &str,
+    ident: OriginPackageIdent,
+    channel: Option<String>,
+) -> Result<OriginPackage> {
+    let token_ref = oauth_token.as_ref().map(|x| &**x);
+    let parent_path = depot.archive_parent(&ident);
+
+    match fs::create_dir_all(parent_path.clone()) {
+        Ok(_) => {}
+        Err(e) => {
+            error!("Unable to create archive directory, err={:?}", e);
+            return Err(Error::IO(e));
+        }
+    };
+
+    // This is pretty frustrating, but I'll explain why this seemingly useless struct
+    // is here. We want to pass None as the last parameter to fetch_package, because
+    // this is running on a server and there's no TTY, so we have no need for
+    // a progress bar. Passing None by itself doesn't work, because the compiler
+    // complains that it can't infer the type for None. Passing None::<DisplayProgress>
+    // doesn't work because the compiler complains that it doesn't know the size of
+    // that parameter. So in order to get this to compile, we need to create this 100%
+    // useless struct that implements the DisplayProgress trait, just so we can cast
+    // None to it when we call the function. If someone has a better way of doing this,
+    // fantastic. I will happily change this code and subscribe to your newsletter.
+    struct NoProgress {}
```

Comment:
Builder has the same issue when making depot client calls - the pattern that was adopted is to create "x_" versions of the calls that don't need to have the progress UI passed in.  If we do that for fetch_package and create a x_fetch_package API (similar to x_download), it should remove the need for the fake progress struct.  Newsletter subscription coming soon.. :)

Reply:
Sadly, these seem like the only options for now - if we go with the no-op struct, we could move it to the depot client as well so that we could take advantage of it from other places as well.

Reply:
Yeah, fair enough.